src/reader/ns_reader.rs

```diff
@@ -0,0 +1,739 @@
+    /// this method without consuming the corresponding [`Start`] event first.
+    ///
+    /// The `end` parameter should contain name of the end element _in the reader
+    /// encoding_. It is good practice to always get that parameter using
+    /// [`BytesStart::to_end()`] method.
+    ///
+    /// # Namespaces
+    ///
+    /// Unlike [`Reader::read_to_end()`], this method resolves namespace
+    /// prefixes, so the names that are not equals literally (for example,
+    /// `a:name` and `b:name`) could be considered equal if prefixes resolved to
+    /// the same namespace.
+    ///
+    /// # Examples
+    ///
+    /// This example shows, how you can skip XML content after you read the
+    /// start event.
+    ///
+    /// ```
+    /// # use pretty_assertions::assert_eq;
+    /// use quick_xml::events::{BytesStart, Event};
+    /// use quick_xml::name::{Namespace, ResolveResult};
+    /// use quick_xml::NsReader;
+    ///
+    /// let mut reader = NsReader::from_str(r#"
+    ///     <outer xmlns="namespace 1">
+    ///         <inner xmlns="namespace 2">
+    ///             <outer></outer>
+    ///         </inner>
+    ///         <inner>
+    ///             <inner></inner>
+    ///             <inner/>
+    ///             <outer></outer>
+    ///             <p:outer xmlns:p="ns"></p:outer>
+    ///             <outer/>
+    ///         </inner>
+    ///     </outer>
+    /// "#);
+    /// reader.trim_text(true);
+    ///
+    /// let ns = Namespace(b"namespace 1");
+    /// let start = BytesStart::borrowed(br#"outer xmlns="namespace 1""#, 5);
+    /// let end   = start.to_end().into_owned();
+    ///
+    /// // First, we read a start event...
+    /// assert_eq!(
+    ///     reader.read_event_ns().unwrap(),
+    ///     (ResolveResult::Bound(ns), Event::Start(start))
+    /// );
+    ///
+    /// //...then, we could skip all events to the corresponding end event.
+    /// // This call will correctly handle nested <outer> elements.
+    /// // Note, however, that this method does not handle namespaces.
+    /// reader.read_to_end(end.name()).unwrap();
+    ///
+    /// // At the end we should get an Eof event, because we ate the whole XML
+    /// assert_eq!(
+    ///     reader.read_event_ns().unwrap(),
+    ///     (ResolveResult::Unbound, Event::Eof)
+    /// );
+    /// ```
+    ///
+    /// [`Start`]: Event::Start
+    /// [`End`]: Event::End
+    /// [`UnexpectedEof`]: crate::errors::Error::UnexpectedEof
+    /// [`read_to_end()`]: Self::read_to_end
+    /// [`BytesStart::to_end()`]: crate::events::BytesStart::to_end
+    #[inline]
+    pub fn read_to_end(&mut self, end: QName) -> Result<()> {
+        // According to the https://www.w3.org/TR/xml11/#dt-etag, end name should
+        // match literally the start name. See `Self::check_end_names` documentation
+        self.reader.read_to_end(end)
+    }
+}
+
+impl<R> From<Reader<R>> for NsReader<R> {
```

Comment:
If the user were to create a normal reader, read some events, then call `NsReader::from(reader)` as this would allow, is that going to result in incorrect internal state?

Reply:
Yes, you are right. It is better to remove it

Reply:
Removed (replaced by private `new` method)